From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Fri, 17 Apr 2020 12:44:15 +0300
Subject: [PATCH 13/67] Allow compositor specializations to override the
 composite (part 2).

Original author: Andrew den Exter <andrew.den.exter@qinetic.com.au>
SHA1: ba21a8694e261769fa53e54a846eeaa651826769

CompositeToTarget now accepts VsyncId as first parameter.
See upstream commit: 011bd0697fc5c1a99e6c3c691aee9b13549d983b

[sailfishos][gecko] Make it possible to extend CompositorBridgeParent

Original author: Raine Makelainen <raine.makelainen@jolla.com>
---
 gfx/layers/ipc/CompositorVsyncScheduler.cpp    | 12 ++++++++----
 gfx/layers/ipc/CompositorVsyncSchedulerOwner.h |  2 ++
 gfx/layers/wr/WebRenderBridgeParent.cpp        |  5 +++++
 gfx/layers/wr/WebRenderBridgeParent.h          |  2 ++
 4 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/gfx/layers/ipc/CompositorVsyncScheduler.cpp b/gfx/layers/ipc/CompositorVsyncScheduler.cpp
index 6ef890301507..0aeaeeff2ba7 100644
--- a/gfx/layers/ipc/CompositorVsyncScheduler.cpp
+++ b/gfx/layers/ipc/CompositorVsyncScheduler.cpp
@@ -267,9 +267,8 @@ void CompositorVsyncScheduler::Composite(const VsyncEvent& aVsyncEvent,
     mCompositeRequestedAt = TimeStamp();
     mLastComposeTime = SampleTime::FromVsync(aVsyncEvent.mTime);
 
-    // Tell the owner to do a composite
-    mVsyncSchedulerOwner->CompositeToTarget(aVsyncEvent.mId, aReasons, nullptr,
-                                            nullptr);
+    // Tell the owner to do a composite.
+    mVsyncSchedulerOwner->CompositeToDefaultTarget(aVsyncEvent.mId, aReasons);
 
     mVsyncNotificationsSkipped = 0;
 
@@ -306,7 +305,12 @@ void CompositorVsyncScheduler::ForceComposeToTarget(wr::RenderReasons aReasons,
 
   mLastComposeTime = SampleTime::FromNow();
   MOZ_ASSERT(mVsyncSchedulerOwner);
-  mVsyncSchedulerOwner->CompositeToTarget(VsyncId(), aReasons, aTarget, aRect);
+  if (aTarget) {
+    mVsyncSchedulerOwner->CompositeToTarget(VsyncId(), aReasons, aTarget,
+                                            aRect);
+  } else {
+    mVsyncSchedulerOwner->CompositeToDefaultTarget(VsyncId(), aReasons);
+  }
 }
 
 bool CompositorVsyncScheduler::NeedsComposite() {
diff --git a/gfx/layers/ipc/CompositorVsyncSchedulerOwner.h b/gfx/layers/ipc/CompositorVsyncSchedulerOwner.h
index ac147fd24159..413c81e10217 100644
--- a/gfx/layers/ipc/CompositorVsyncSchedulerOwner.h
+++ b/gfx/layers/ipc/CompositorVsyncSchedulerOwner.h
@@ -26,6 +26,8 @@ class CompositorVsyncSchedulerOwner {
                                  gfx::DrawTarget* aTarget,
                                  const gfx::IntRect* aRect = nullptr) = 0;
   virtual TimeDuration GetVsyncInterval() const = 0;
+  virtual void CompositeToDefaultTarget(VsyncId aId,
+                                        wr::RenderReasons aReasons) = 0;
 };
 
 }  // namespace layers
diff --git a/gfx/layers/wr/WebRenderBridgeParent.cpp b/gfx/layers/wr/WebRenderBridgeParent.cpp
index 3548116af06b..70834b964fe9 100644
--- a/gfx/layers/wr/WebRenderBridgeParent.cpp
+++ b/gfx/layers/wr/WebRenderBridgeParent.cpp
@@ -2923,5 +2923,10 @@ WebRenderBridgeParentRef::~WebRenderBridgeParentRef() {
   MOZ_ASSERT(!mWebRenderBridge);
 }
 
+void WebRenderBridgeParent::CompositeToDefaultTarget(
+    VsyncId aId, wr::RenderReasons aReasons) {
+  CompositeToTarget(aId, aReasons, nullptr, nullptr);
+}
+
 }  // namespace layers
 }  // namespace mozilla
diff --git a/gfx/layers/wr/WebRenderBridgeParent.h b/gfx/layers/wr/WebRenderBridgeParent.h
index 308a829e84e2..69706f42fdf1 100644
--- a/gfx/layers/wr/WebRenderBridgeParent.h
+++ b/gfx/layers/wr/WebRenderBridgeParent.h
@@ -189,6 +189,8 @@ class WebRenderBridgeParent final : public PWebRenderBridgeParent,
   void CompositeToTarget(VsyncId aId, wr::RenderReasons aReasons,
                          gfx::DrawTarget* aTarget,
                          const gfx::IntRect* aRect = nullptr) override;
+  void CompositeToDefaultTarget(VsyncId aId,
+                                wr::RenderReasons aReasons) override;
   TimeDuration GetVsyncInterval() const override;
 
   // CompositableParentManager
-- 
2.51.0

